### DISCUSSION TITLE
Mass Balance PorousFlow

### DISCUSSION THREAD

**Discussion by JonhWang55** [idaholab/moose#17530] [Q&A Modules: General] [answered]
Dear all,

Firstly, a big thank to the MOOSE community for the this incredible tool. I appreciate the complete documentation, in particular PorousFlow's which is quite good. 

I'm running a medium-size simulation with PorousFlow that includes different types of BCs and sinks. It works really well (despite it took me a while to make it converge). Nonetheless, while reading my results I found an odd behavior in the mass balance. To illustrate this, I have set a simple input file. The system has a PorousFlowPiecewiseLinearSink on the left side which increases the porepressure. The simulation runs for ten seconds while the fluid flux and water balance are computed in post with NodalSum and PorousFlowFluidMass respectively. Now, I would expect that the total fluid going into the system (computed as fluid flux times timestep) would match the fluid balance. But it looks like very different results are obtained (as you may see if you run the simulation). Most likely is a units confusion but I can't figure where is it. 

https://drive.google.com/file/d/1quHSxTuHVNsq_um5lmEkEqWxtlONMFN1/view?usp=sharing

Thanks!
Jhon

**Comment by WilkAndy** [comment, marked as answer]
You have discovered a weakness in [PorousFlowBasicTHM](https://mooseframework.inl.gov/source/actions/PorousFlowBasicTHM.html) : it does not play well with the Sinks or PorousFlowFluidMass.  While you can reconcile everything (remember BasicTHM works with fluid volumes, not fluid masses), if you are using Sinks and wanting to look at mass balances, i strongly recommend using [PorousFlowFullySaturated](https://mooseframework.inl.gov/source/actions/PorousFlowFullySaturated.html).  Then you don't have to think!  All you need to is change `PorousFlowBasicTHM` to `PorousFlowFullySaturated` in your input file.  Debug your model using FullySaturated, and then maybe go back to BasicTHM for production runs if you need ultimate speed.

**Reply by WilkAndy** [reply]
I should have added: if you want to continue with `PorousFowBasicTHM` then it's probably easiest to set
```
  multiply_by_density = true
```
in the `PorousFlowBasicTHM` block.  Then you don't have to do so much thinking :-).  Secondly, remember to set your fluid bulk modulus consistently throughout the input file.  `SimpleFluidProperties` defaults to 2E9, so in the `PorousFlowConstantBiotModulus` you should set
```
    fluid_bulk_modulus = 2E9
```
These two changes will give you the result you want (with the disclaimers about constant Biot modulus that you can find discussed [here](https://mooseframework.inl.gov/source/kernels/PorousFlowFullySaturatedMassTimeDerivative.html) ).

**Reply by JonhWang55** [reply]
Dear Andy,

Thanks! Both, changing from `PorousFlowBasicTHM` to `PorousFlowFullySaturated` and activating `multiply_by_density = true` closed  the mass balance correctly.

Overall: 
- Changing `PorousFlowBasicTHM` works with fluid volumes and activating the flag `multiply_by_density = true` will change that to mass.

- `PorousFlowFullySaturated` works with mass fluxes, however, is not as fast as `PorousFlowBasicTHM`. 

Thanks!
